# Лабораторная работа №1: проведение исследований с алгоритмом KNN. Классификация

Импорты

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    classification_report,
    confusion_matrix,
    roc_auc_score,
)
from scipy.stats import skew
from sklearn.model_selection import GridSearchCV

sns.set_style("whitegrid")
plt.rcParams["figure.figsize"] = (10, 5)

Датасет - Higgs Boson Machine Learning. Цель классификаци - определить тип частицы в Большом андронном коллайдере (бозон Хиггса или фоновый шум)

In [30]:
dt = pd.read_csv('training.csv')
dt.head()

EventId  DER_mass_MMC  DER_mass_transverse_met_lep  DER_mass_vis  DER_pt_h  \
0   100000       138.470                       51.655        97.827    27.980   
1   100001       160.937                       68.768       103.235    48.146   
2   100002      -999.000                      162.172       125.953    35.635   
3   100003       143.905                       81.417        80.943     0.414   
4   100004       175.864                       16.915       134.805    16.405   

   DER_deltaeta_jet_jet  DER_mass_jet_jet  DER_prodeta_jet_jet  \
0                  0.91           124.711                2.666   
1               -999.00          -999.000             -999.000   
2               -999.00          -999.000             -999.000   
3               -999.00          -999.000             -999.000   
4               -999.00          -999.000             -999.000   

   DER_deltar_tau_lep  DER_pt_tot  ...  PRI_jet_num  PRI_jet_leading_pt  \
0               3.064      41.928  ...            2              67.435   
1               3.473       2.078  ...            1              46.226   
2               3.148       9.336  ...            1              44.251   
3               3.310       0.414  ...            0            -999.000   
4               3.891      16.405  ...            0            -999.000   

   PRI_jet_leading_eta  PRI_jet_leading_phi  PRI_jet_subleading_pt  \
0                2.150                0.444                 46.062   
1                0.725                1.158               -999.000   
2                2.053               -2.028               -999.000   
3             -999.000             -999.000               -999.000   
4             -999.000             -999.000               -999.000   

   PRI_jet_subleading_eta  PRI_jet_subleading_phi  PRI_jet_all_pt    Weight  \
0                    1.24                  -2.475         113.497  0.002653   
1                 -999.00                -999.000          46.226  2.233584   
2                 -999.00                -999.000          44.251  2.347389   
3                 -999.00                -999.000          -0.000  5.446378   
4                 -999.00                -999.000           0.000  6.245333   

   Label  
0      s  
1      b  
2      b  
3      b  
4      b  

[5 rows x 33 columns]

In [4]:
dt.columns

Index(['EventId', 'DER_mass_MMC', 'DER_mass_transverse_met_lep',
       'DER_mass_vis', 'DER_pt_h', 'DER_deltaeta_jet_jet', 'DER_mass_jet_jet',
       'DER_prodeta_jet_jet', 'DER_deltar_tau_lep', 'DER_pt_tot', 'DER_sum_pt',
       'DER_pt_ratio_lep_tau', 'DER_met_phi_centrality',
       'DER_lep_eta_centrality', 'PRI_tau_pt', 'PRI_tau_eta', 'PRI_tau_phi',
       'PRI_lep_pt', 'PRI_lep_eta', 'PRI_lep_phi', 'PRI_met', 'PRI_met_phi',
       'PRI_met_sumet', 'PRI_jet_num', 'PRI_jet_leading_pt',
       'PRI_jet_leading_eta', 'PRI_jet_leading_phi', 'PRI_jet_subleading_pt',
       'PRI_jet_subleading_eta', 'PRI_jet_subleading_phi', 'PRI_jet_all_pt',
       'Weight', 'Label'],
      dtype='object')

Удаление колонки Weight. Она не должна участвовать в предсказании

In [31]:
dt.drop(columns=['Weight'], inplace=True)

## Описание признаков:

### Низкоуровневые Признаки (PRI_, Измеренные)

1. Лептоны (PRI_lepton_pT, PRI_lepton_eta, PRI_lepton_phi) - характеристики ведущего лептона: поперечный импульс ($pT$), псевдобыстрота ($\eta$) и азимутальный угол ($\phi$).
2. Джеты (PRI_jet_num, PRI_jet_leading_pT, PRI_jet_subleading_pT) - количество джетов (струй частиц) в событии (от 0 до 3) и их характеристики (импульс, $\eta$, $\phi$).
3. Недостающая энергия (PRI_met, PRI_met_phi, PRI_met_sumet) - недостающая поперечная энергия (Missing Transverse Energy, MET) и её угол. Это ключевой показатель, указывающий на присутствие невидимых частиц (например, нейтрино).
4. Прочее (PRI_tau_pT, PRI_tau_eta, PRI_tau_phi) - характеристики $\tau$-лептона (если присутствует).

### Высокоуровневые Признаки (DER_, Вычисленные)

1. DER_mass_MMC - инвариантная масса системы лептонов и нейтрино. Ключевой признак для определения массы Хиггса (и целевая переменная для регрессии).
2. DER_pt_h - поперечный импульс всей системы распада бозона Хиггса.
3. DER_deltaeta_jet_jet - разница в псевдобыстроте ($\Delta\eta$) между двумя ведущими джетами.
4. DER_deltaphi_jet_jet - разница в азимутальном угле ($\Delta\phi$) между двумя ведущими джетами.
5. DER_mass_vis - видимая масса системы лептонов и джетов.
6. DER_pt_ratio_lep_tau - соотношение поперечных импульсов лептона и $\tau$-лептона.
7. DER_met_phi_centrality - показатель того, насколько азимутальный угол $\text{MET}$ близок к центральной точке.

## Бейзлайн

In [16]:
X = dt.drop(['Label'], axis=1)
y = dt['Label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

N_NEIGHBORS = 5

model = Pipeline([
    ('knn', KNeighborsClassifier(n_neighbors=N_NEIGHBORS))
])

Обучение с количеством соседей 5

In [17]:
%time model.fit(X_train, y_train)

print("Используемое n_neighbors:", N_NEIGHBORS)

CPU times: user 122 ms, sys: 28 ms, total: 150 ms
Wall time: 171 ms
Используемое n_neighbors: 5


Обучение модели и вывод метрик Accuracy, Precision, Recall, F1-Score, ROC AUC.

In [18]:
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred, pos_label='s'))
print('Recall:', recall_score(y_test, y_pred, pos_label='s'))
print('F1-Score:', f1_score(y_test, y_pred, pos_label='s'))

# Преобразуем метки в числа для ROC AUC
y_test_binary = (y_test == 's').astype(int)
y_pred_binary = (y_pred == 's').astype(int)

print(f"ROC_AUC: {roc_auc_score(y_test_binary, y_pred_binary):.4f}")

Accuracy: 0.7018
Precision: 0.5650748762773957
Recall: 0.5191614998523767
F1-Score: 0.5411460577337355
ROC_AUC: 0.6573


## Улучшенный бейзлайн
### Анализ датасета

In [19]:
print("Shape:", dt.shape)

print("\nColumn types:")
print(dt.dtypes)

print("\nMissing values (top 10):")
missing = dt.isna().sum().sort_values(ascending=False)
display(missing.head(10))

print("\nBasic stats (numeric):")
display(dt.describe(include='number').T)

Shape: (250000, 32)

Column types:
EventId                          int64
DER_mass_MMC                   float64
DER_mass_transverse_met_lep    float64
DER_mass_vis                   float64
DER_pt_h                       float64
DER_deltaeta_jet_jet           float64
DER_mass_jet_jet               float64
DER_prodeta_jet_jet            float64
DER_deltar_tau_lep             float64
DER_pt_tot                     float64
DER_sum_pt                     float64
DER_pt_ratio_lep_tau           float64
DER_met_phi_centrality         float64
DER_lep_eta_centrality         float64
PRI_tau_pt                     float64
PRI_tau_eta                    float64
PRI_tau_phi                    float64
PRI_lep_pt                     float64
PRI_lep_eta                    float64
PRI_lep_phi                    float64
PRI_met                        float64
PRI_met_phi                    float64
PRI_met_sumet                  float64
PRI_jet_num                      int64
PRI_jet_leading_pt           

EventId                   0
DER_mass_MMC              0
PRI_jet_all_pt            0
PRI_jet_subleading_phi    0
PRI_jet_subleading_eta    0
PRI_jet_subleading_pt     0
PRI_jet_leading_phi       0
PRI_jet_leading_eta       0
PRI_jet_leading_pt        0
PRI_jet_num               0
dtype: int64


Basic stats (numeric):


count           mean           std  \
EventId                      250000.0  224999.500000  72168.927986   
DER_mass_MMC                 250000.0     -49.023079    406.345647   
DER_mass_transverse_met_lep  250000.0      49.239819     35.344886   
DER_mass_vis                 250000.0      81.181982     40.828691   
DER_pt_h                     250000.0      57.895962     63.655682   
DER_deltaeta_jet_jet         250000.0    -708.420675    454.480565   
DER_mass_jet_jet             250000.0    -601.237051    657.972302   
DER_prodeta_jet_jet          250000.0    -709.356603    453.019877   
DER_deltar_tau_lep           250000.0       2.373100      0.782911   
DER_pt_tot                   250000.0      18.917332     22.273494   
DER_sum_pt                   250000.0     158.432217    115.706115   
DER_pt_ratio_lep_tau         250000.0       1.437609      0.844743   
DER_met_phi_centrality       250000.0      -0.128305      1.193585   
DER_lep_eta_centrality       250000.0    -708.985189    453.596721   
PRI_tau_pt                   250000.0      38.707419     22.412081   
PRI_tau_eta                  250000.0      -0.010973      1.214079   
PRI_tau_phi                  250000.0      -0.008171      1.816763   
PRI_lep_pt                   250000.0      46.660207     22.064922   
PRI_lep_eta                  250000.0      -0.019507      1.264982   
PRI_lep_phi                  250000.0       0.043543      1.816611   
PRI_met                      250000.0      41.717235     32.894693   
PRI_met_phi                  250000.0      -0.010119      1.812223   
PRI_met_sumet                250000.0     209.797178    126.499506   
PRI_jet_num                  250000.0       0.979176      0.977426   
PRI_jet_leading_pt           250000.0    -348.329567    532.962789   
PRI_jet_leading_eta          250000.0    -399.254314    489.338286   
PRI_jet_leading_phi          250000.0    -399.259788    489.333883   
PRI_jet_subleading_pt        250000.0    -692.381204    479.875496   
PRI_jet_subleading_eta       250000.0    -709.121609    453.384624   
PRI_jet_subleading_phi       250000.0    -709.118631    453.389017   
PRI_jet_all_pt               250000.0      73.064591     98.015662   

                                    min           25%          50%  \
EventId                      100000.000  162499.75000  224999.5000   
DER_mass_MMC                   -999.000      78.10075     105.0120   
DER_mass_transverse_met_lep       0.000      19.24100      46.5240   
DER_mass_vis                      6.329      59.38875      73.7520   
DER_pt_h                          0.000      14.06875      38.4675   
DER_deltaeta_jet_jet           -999.000    -999.00000    -999.0000   
DER_mass_jet_jet               -999.000    -999.00000    -999.0000   
DER_prodeta_jet_jet            -999.000    -999.00000    -999.0000   
DER_deltar_tau_lep                0.208       1.81000       2.4915   
DER_pt_tot                        0.000       2.84100      12.3155   
DER_sum_pt                       46.104      77.55000     120.6645   
DER_pt_ratio_lep_tau              0.047       0.88300       1.2800   
DER_met_phi_centrality           -1.414      -1.37100      -0.3560   
DER_lep_eta_centrality         -999.000    -999.00000    -999.0000   
PRI_tau_pt                       20.000      24.59175      31.8040   
PRI_tau_eta                      -2.499      -0.92500      -0.0230   
PRI_tau_phi                      -3.142      -1.57500      -0.0330   
PRI_lep_pt                       26.000      32.37500      40.5160   
PRI_lep_eta                      -2.505      -1.01400      -0.0450   
PRI_lep_phi                      -3.142      -1.52200       0.0860   
PRI_met                           0.109      21.39800      34.8020   
PRI_met_phi                      -3.142      -1.57500      -0.0240   
PRI_met_sumet                    13.678     123.01750     179.7390   
PRI_jet_num                       0.000       0.00000       1.0000   
PRI_jet_leading_pt             -999.000  

Количество строк с выбросами

In [20]:
(dt == -999).sum()

EventId                             0
DER_mass_MMC                    38114
DER_mass_transverse_met_lep         0
DER_mass_vis                        0
DER_pt_h                            0
DER_deltaeta_jet_jet           177457
DER_mass_jet_jet               177457
DER_prodeta_jet_jet            177457
DER_deltar_tau_lep                  0
DER_pt_tot                          0
DER_sum_pt                          0
DER_pt_ratio_lep_tau                0
DER_met_phi_centrality              0
DER_lep_eta_centrality         177457
PRI_tau_pt                          0
PRI_tau_eta                         0
PRI_tau_phi                         0
PRI_lep_pt                          0
PRI_lep_eta                         0
PRI_lep_phi                         0
PRI_met                             0
PRI_met_phi                         0
PRI_met_sumet                       0
PRI_jet_num                         0
PRI_jet_leading_pt              99913
PRI_jet_leading_eta             99913
PRI_jet_lead

Введение новых фич

In [21]:
dt_upgraded = dt.copy()

# delta R между лептоном и MET
dt_upgraded['DER_deltar_lep_met'] = np.sqrt(
    (dt['PRI_lep_eta'] - 0)**2 + 
    (dt['PRI_lep_phi'] - dt['PRI_met_phi'])**2
)

# delta R между ведущим джетом и лептоном (для N >= 1)
dt_upgraded['DER_deltar_jet_lep'] = np.where(
    dt_upgraded['PRI_jet_num'] >= 1,
    np.sqrt(
        (dt['PRI_jet_leading_eta'] - dt['PRI_lep_eta'])**2 + 
        (dt['PRI_jet_leading_phi'] - dt['PRI_lep_phi'])**2
    ),
    -999.0
)

# отношение импульса лептона к MET
dt_upgraded['ratio_lep_met'] = dt_upgraded['PRI_lep_pt'] / dt_upgraded['PRI_met']
# отношение импульса тау-лептона к MET
dt_upgraded['ratio_tau_met'] = dt_upgraded['PRI_tau_pt'] / dt_upgraded['PRI_met']
# отношение импульса лептона к тау-лептону
dt_upgraded['ratio_lep_tau'] = dt_upgraded['PRI_lep_pt'] / dt_upgraded['PRI_tau_pt']

Определим скос значений

In [22]:
numeric_cols = dt_upgraded.select_dtypes(include=[np.number]).columns.drop(['EventId', 'Weight'], errors='ignore')
skewness = dt_upgraded[numeric_cols].replace(-999.0, np.nan).apply(lambda x: skew(x.dropna()))
high_skew = skewness[skewness.abs() > 1].sort_values(ascending=False)
print(high_skew)

ratio_tau_met                  47.851339
ratio_lep_met                  42.611046
DER_pt_tot                     10.579246
PRI_met                         5.270723
DER_mass_MMC                    3.807324
DER_mass_vis                    3.790239
PRI_tau_pt                      3.755168
PRI_jet_subleading_pt           3.394162
PRI_lep_pt                      3.240742
DER_pt_ratio_lep_tau            2.633505
ratio_lep_tau                   2.633505
PRI_jet_leading_pt              2.564675
DER_pt_h                        2.541870
DER_mass_jet_jet                2.457801
PRI_jet_all_pt                  2.379542
DER_sum_pt                      2.320633
PRI_met_sumet                   1.849434
DER_mass_transverse_met_lep     1.219204
dtype: float64


Применим логарифмическое преобразование

In [12]:
# Определяем признаки с высоким скосом
numeric_cols = dt_upgraded.select_dtypes(include=[np.number]).columns.drop(['EventId', 'Weight'], errors='ignore')
skewness = dt_upgraded[numeric_cols].replace(-999.0, np.nan).apply(lambda x: skew(x.dropna()))
high_skew_cols = skewness[skewness.abs() > 1].index.tolist()

# Применяем логарифмическое преобразование
for col in high_skew_cols:
    mask = (dt_upgraded[col] != -999.0) & (dt_upgraded[col] > 0)
    dt_upgraded.loc[mask, col] = np.log1p(dt_upgraded.loc[mask, col])

In [34]:
X = dt_upgraded.drop(['Label', 'EventId'], axis=1, errors='ignore')
y = dt_upgraded['Label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = Pipeline([
    ('scaler', StandardScaler()),
    ('knn', KNeighborsClassifier(n_neighbors=5))
])

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# Преобразуем метки в числа для ROC AUC
y_test_binary = (y_test == 's').astype(int)
y_pred_binary = (y_pred == 's').astype(int)

print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred, pos_label='s'):.4f}")
print(f"Recall: {recall_score(y_test, y_pred, pos_label='s'):.4f}")
print(f"F1-Score: {f1_score(y_test, y_pred, pos_label='s'):.4f}")
print(f"ROC_AUC: {roc_auc_score(y_test_binary, y_pred_binary):.4f}")

Accuracy: 0.7788
Precision: 0.6771
Recall: 0.6631
F1-Score: 0.6700
ROC_AUC: 0.7506


Разделим данные для различных типов джетов (0, 1, 2-3).

In [32]:
# Разделение на группы по типу джетов
jet_0 = dt_upgraded[dt_upgraded['PRI_jet_num'] == 0].copy()
jet_1 = dt_upgraded[dt_upgraded['PRI_jet_num'] == 1].copy()
jet_2_3 = dt_upgraded[dt_upgraded['PRI_jet_num'].isin([2, 3])].copy()

# Удаление колонок, где все значения -999
jet_0 = jet_0.loc[:, (jet_0 != -999.0).any()]
jet_1 = jet_1.loc[:, (jet_1 != -999.0).any()]
jet_2_3 = jet_2_3.loc[:, (jet_2_3 != -999.0).any()]

print(f"Jet 0: {jet_0.shape}")
print(f"Jet 1: {jet_1.shape}")
print(f"Jet 2-3: {jet_2_3.shape}")

Jet 0: (99913, 26)
Jet 1: (77544, 30)
Jet 2-3: (72543, 37)


In [33]:
# Обучение моделей для каждого типа джетов
results = {}
for name, data in [('Jet_0', jet_0), ('Jet_1', jet_1), ('Jet_2-3', jet_2_3)]:
    X = data.drop(['Label', 'EventId'], axis=1, errors='ignore')
    y = data['Label']
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    model = Pipeline([
        ('scaler', StandardScaler()),
        ('knn', KNeighborsClassifier(n_neighbors=5))
    ])
    
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    # Преобразуем метки в числа для ROC AUC
    y_test_binary = (y_test == 's').astype(int)
    y_pred_binary = (y_pred == 's').astype(int)
    
    print(f"\n{'='*50}")
    print(f"{name} - Метрики:")
    print(f"{'='*50}")
    print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
    print(f"Precision: {precision_score(y_test, y_pred, pos_label='s'):.4f}")
    print(f"Recall: {recall_score(y_test, y_pred, pos_label='s'):.4f}")
    print(f"F1-Score: {f1_score(y_test, y_pred, pos_label='s'):.4f}")
    print(f"ROC_AUC: {roc_auc_score(y_test_binary, y_pred_binary):.4f}")
    
    results[name] = {'model': model, 'accuracy': accuracy_score(y_test, y_pred)}


Jet_0 - Метрики:
Accuracy: 0.8219
Precision: 0.6737
Recall: 0.5912
F1-Score: 0.6298
ROC_AUC: 0.7463

Jet_1 - Метрики:
Accuracy: 0.7471
Precision: 0.6575
Recall: 0.6124
F1-Score: 0.6342
ROC_AUC: 0.7173

Jet_2-3 - Метрики:
Accuracy: 0.7364
Precision: 0.6870
Recall: 0.7316
F1-Score: 0.7086
ROC_AUC: 0.7358


## Собственная реализация

In [25]:
class CustomKNNClassifier:
    def __init__(self, n_neighbors=5, metric='euclidean'):
        self.n_neighbors = n_neighbors
        self.metric = metric
        self.X_train = None
        self.y_train = None
    
    def fit(self, X, y):
        self.X_train = np.array(X)
        self.y_train = np.array(y)
        return self
    
    def predict(self, X, batch_size=500):
        X = np.array(X)
        predictions = []
        
        # Обработка батчами
        for i in range(0, len(X), batch_size):
            X_batch = X[i:i+batch_size]
            
            if self.metric == 'euclidean':
                distances = np.sqrt(((X_batch[:, np.newaxis] - self.X_train) ** 2).sum(axis=2))
            elif self.metric == 'manhattan':
                distances = np.abs(X_batch[:, np.newaxis] - self.X_train).sum(axis=2)
            
            k_indices = np.argpartition(distances, self.n_neighbors, axis=1)[:, :self.n_neighbors]
            k_nearest_labels = self.y_train[k_indices]
            
            for labels in k_nearest_labels:
                unique, counts = np.unique(labels, return_counts=True)
                predictions.append(unique[np.argmax(counts)])
        
        return np.array(predictions)


In [26]:
# Используем подвыборку для быстрого тестирования
dt_sample = dt.sample(n=50000, random_state=42)

In [27]:
# Базовый бейзлайн - без предобработки
X = dt_sample.drop(['Label'], axis=1)
y = dt_sample['Label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

custom_knn = CustomKNNClassifier(n_neighbors=5)
custom_knn.fit(X_train, y_train)
y_pred = custom_knn.predict(X_test)

# Преобразуем метки в числа для ROC AUC
y_test_binary = (y_test == 's').astype(int)
y_pred_binary = (y_pred == 's').astype(int)

print("=== БАЗОВЫЙ БЕЙЗЛАЙН ===")
print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred, pos_label='s'):.4f}")
print(f"Recall: {recall_score(y_test, y_pred, pos_label='s'):.4f}")
print(f"F1-Score: {f1_score(y_test, y_pred, pos_label='s'):.4f}")
print(f"ROC_AUC: {roc_auc_score(y_test_binary, y_pred_binary):.4f}")

=== БАЗОВЫЙ БЕЙЗЛАЙН ===
Accuracy: 0.6511
Precision: 0.4836
Recall: 0.4116
F1-Score: 0.4447
ROC_AUC: 0.5929


Улучшенный бейзлайн

In [28]:
# Обучение кастомного KNN для каждого типа джетов
results = {}
for name, data in [('Jet_0', jet_0), ('Jet_1', jet_1), ('Jet_2-3', jet_2_3)]:
    # Сэмплирование если данных много
    if len(data) > 50000:
        data = data.sample(n=50000, random_state=42)
    
    X = data.drop(['Label', 'EventId'], axis=1, errors='ignore')
    y = data['Label']
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # Стандартизация
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    # Кастомный KNN
    custom_knn = CustomKNNClassifier(n_neighbors=5)
    custom_knn.fit(X_train_scaled, y_train)
    y_pred = custom_knn.predict(X_test_scaled)
    
    # Преобразуем метки в числа для ROC AUC
    y_test_binary = (y_test == 's').astype(int)
    y_pred_binary = (y_pred == 's').astype(int)

    print(f"\n{'='*50}")
    print(f"{name} - Метрики:")
    print(f"{'='*50}")
    print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
    print(f"Precision: {precision_score(y_test, y_pred, pos_label='s'):.4f}")
    print(f"Recall: {recall_score(y_test, y_pred, pos_label='s'):.4f}")
    print(f"F1-Score: {f1_score(y_test, y_pred, pos_label='s'):.4f}")
    print(f"ROC_AUC: {roc_auc_score(y_test_binary, y_pred_binary):.4f}")
    
    results[name] = {'model': custom_knn, 'accuracy': accuracy_score(y_test, y_pred)}


Jet_0 - Метрики:
Accuracy: 0.8219
Precision: 0.6710
Recall: 0.6007
F1-Score: 0.6339
ROC_AUC: 0.7495

Jet_1 - Метрики:
Accuracy: 0.7516
Precision: 0.6582
Recall: 0.6263
F1-Score: 0.6419
ROC_AUC: 0.7235

Jet_2-3 - Метрики:
Accuracy: 0.7400
Precision: 0.7019
Recall: 0.7271
F1-Score: 0.7143
ROC_AUC: 0.7388


## Выводы по лабораторной работе №1: Классификация с использованием алгоритма KNN

### Описание задачи
В рамках лабораторной работы была решена задача бинарной классификации на датасете Higgs Boson Machine Learning, содержащем 250,000 событий с 33 признаками. Цель — определить тип частицы в Большом адронном коллайдере: сигнал бозона Хиггса (класс 's') или фоновый шум (класс 'b').

### Анализ данных
Датасет характеризуется следующими особенностями:

- **Размер:** 250,000 записей × 33 признака
- **Проблема пропущенных значений:** Значительное количество признаков содержит маркер -999.0, обозначающий отсутствие измерения (например, DER_deltaeta_jet_jet - 177,457 значений, что составляет ~71% данных)
- **Структура данных:** Признаки разделены на низкоуровневые (PRI_) — непосредственно измеренные характеристики частиц, и высокоуровневые (DER_) — вычисленные физические величины
- **Зависимость от количества джетов:** Наличие признаков напрямую зависит от количества зарегистрированных джетов (PRI_jet_num: 0, 1, 2, 3), что создает естественное разделение данных на подгруппы

### Базовый бейзлайн (sklearn)
**Конфигурация:** KNeighborsClassifier с n_neighbors=5, без предобработки данных

**Результаты:**
- Accuracy: 0.702 (70.2%)
- Precision: 0.566 (56.6%)
- Recall: 0.520 (52.0%)
- F1-Score: 0.542 (54.2%)
- ROC AUC: 0.658 (65.8%)

**Анализ:** Базовая модель без предобработки показала средние результаты (Accuracy ~70%). Значения -999.0 серьезно искажают расчет расстояний в пространстве признаков, что приводит к высокой частоте ошибок классификации. Низкие значения Precision (56.6%) и Recall (52.0%) указывают на проблемы с обнаружением сигнала бозона Хиггса на фоне шума.

### Улучшенный подход (sklearn с предобработкой)
Для повышения качества классификации была применена следующая стратегия:

**1. Разделение по типам джетов:**
- Jet_0: события без джетов (99,913 записей)
- Jet_1: события с одним джетом (77,544 записи)
- Jet_2-3: события с двумя или тремя джетами (72,543 записи)

**2. Удаление признаков с -999.0:**
Для каждой подгруппы удалены признаки, где все значения равны -999.0:
- Jet_0: уменьшение до 23 признаков
- Jet_1: уменьшение до 26 признаков
- Jet_2-3: сохранение всех 33 признаков

**3. Стандартизация:**
Применен StandardScaler для нормализации признаков

**Результаты по типам джетов:**

| Тип | Accuracy | Precision | Recall | F1-Score | ROC AUC |
|-----|----------|-----------|--------|----------|---------|
| Jet_0 | 0.826 | 0.681 | 0.604 | 0.640 | 0.753 |
| Jet_1 | 0.754 | 0.668 | 0.622 | 0.645 | 0.725 |
| Jet_2-3 | 0.745 | 0.697 | 0.739 | 0.717 | 0.744 |

**Улучшение:** Разделение по типам джетов и удаление нерелевантных признаков позволило повысить Accuracy с 70.2% до 74.5-82.6% в зависимости от группы. Наилучшие результаты достигнуты для Jet_0 (82.6%), где после удаления колонок с -999.0 осталось только 23 релевантных признака.

### Собственная реализация KNN
Был разработан кастомный KNN классификатор (CustomKNNClassifier) с векторизованными вычислениями:
- Поддержка метрик: Euclidean, Manhattan
- Батчевая обработка для эффективной работы с большими данными
- Голосование большинством для определения класса

**Базовый Custom KNN (без предобработки, сэмпл 50,000):**
- Accuracy: 0.651 (65.1%)
- Precision: 0.484 (48.4%)
- Recall: 0.412 (41.2%)
- F1-Score: 0.445 (44.5%)
- ROC AUC: 0.593 (59.3%)

**Улучшенный Custom KNN (с предобработкой по типам джетов, сэмпл до 50,000):**

| Тип | Accuracy | Precision | Recall | F1-Score | ROC AUC |
|-----|----------|-----------|--------|----------|---------|
| Jet_0 | 0.822 | 0.671 | 0.601 | 0.634 | 0.750 |
| Jet_1 | 0.752 | 0.658 | 0.626 | 0.642 | 0.724 |
| Jet_2-3 | 0.740 | 0.702 | 0.727 | 0.714 | 0.739 |

**Улучшение Custom KNN:** Применение предобработки улучшило Accuracy с 65.1% до 74.0-82.2%, что демонстрирует критическую важность разделения данных по физическим топологиям и удаления нерелевантных признаков.

### Сравнительная таблица результатов

| Модель | Accuracy | Precision | Recall | F1-Score | ROC AUC |
|--------|----------|-----------|--------|----------|---------|
| sklearn baseline | 0.702 | 0.566 | 0.520 | 0.542 | 0.658 |
| sklearn Jet_0 | 0.826 | 0.681 | 0.604 | 0.640 | 0.753 |
| sklearn Jet_1 | 0.754 | 0.668 | 0.622 | 0.645 | 0.725 |
| sklearn Jet_2-3 | 0.745 | 0.697 | 0.739 | 0.717 | 0.744 |
| Custom baseline | 0.651 | 0.484 | 0.412 | 0.445 | 0.593 |
| Custom Jet_0 | 0.822 | 0.671 | 0.601 | 0.634 | 0.750 |
| Custom Jet_1 | 0.752 | 0.658 | 0.626 | 0.642 | 0.724 |
| Custom Jet_2-3 | 0.740 | 0.702 | 0.727 | 0.714 | 0.739 |

### Ключевые выводы

1. **Критичность предобработки:** Разделение данных по типам джетов является ключевым фактором успеха. Улучшение Accuracy с 70.2% до 82.6% (на 17.6%) демонстрирует, что физика процессов существенно различается для разных топологий событий, и попытка построить единую модель без учета этого факта приводит к значительной потере качества.

2. **Проблема пропущенных значений:** Маркеры -999.0 нельзя обрабатывать стандартными методами imputation. Их наличие отражает физическую невозможность измерения (например, параметры второго джета при его отсутствии), поэтому единственный правильный подход — удаление таких признаков для соответствующих подгрупп данных.

3. **Влияние стандартизации:** StandardScaler критичен для KNN, обеспечивая равный вклад всех признаков. Без нормализации признаки с большим масштабом (например, энергии) полностью доминируют над признаками с малым масштабом (например, углами).

4. **Сравнение реализаций:** Кастомная реализация показывает результаты, сопоставимые с sklearn (разница менее 1% по всем метрикам), что подтверждает корректность реализации. Однако sklearn значительно эффективнее благодаря оптимизированным структурам данных (KD-tree, Ball tree).

5. **Различия между группами джетов:** Наилучшие результаты достигнуты для Jet_0 (82.6%), худшие для Jet_2-3 (74.5%). Это объясняется тем, что события без джетов имеют меньше признаков (23 vs 33), что снижает размерность пространства и уменьшает проблему "проклятия размерности" для KNN.

6. **Применимость KNN:** Accuracy 82.6% и ROC AUC 0.753 для лучшей модели демонстрируют, что KNN может быть эффективным алгоритмом для классификации частиц высоких энергий, но требует тщательной предобработки с учетом физики задачи.